packages/@aws-cdk/cdk/lib/core/tokens.ts

```diff
@@ -29,6 +31,43 @@ export class Token {
 
         return value;
     }
+
+    /**
+     * Return a string representation of this token
+     *
+     * This string representation can be embedded into strings and
+     * will be turned into the actual representation of the token
+     * when resolve() is called on the string.
+     */
+    public toString(): string {
+        if (this.stringRepr === undefined) {
+            this.stringRepr = TOKEN_STRING_MAP.register(this);
+        }
+        return this.stringRepr;
+    }
+
+    /**
+     * Return the JSON representation of this Token
+     *
+     * Implements JSON.stringify()-compatibility for this Token. Note that it
+     * returns a string representation for the Token pre-resolution, NOT the
+     * resolved value of the Token.
+     */
+    public toJSON(): any {
```

Comment:
Why is this needed?

Reply:
This gets called as custom serialization in `JSON.stringify()`.

Reply:
Wow! That's cool. Good to know. 

So now I am thinking if we can simplify `resolve` by simply saying that tokens are objects that implement `toJSON` instead of `resolve`, but then why are you returning the string representation here....

Reply:
`JSON.stringify()` must always return a string.

Structural reason: if we returned a CloudFormation intrinsic here, it might work if our `Token` is the top-level object, but it might also be 3 levels deep... but we still want to wrap the WHOLE return value in a `FnConcat`. That cannot be done in the `toJSON`, because that only has a local view of what it's producing. So we must first return a string that nicely embeds into the larger JSONified string, and only later split that string up again and wrap it in a `FnConcat` at a later point when we're calling `resolve()` on the entire string.

And also the simple reason: API doesn't allow it. Whatever `toJSON` returns is stringified according to the normal stringification rules. So if we returned an object here, we would end up with `"{ Fn::GetAtt: [...] }"` inside our JSON string, not in a position where CloudFormation would evaluate it.

Reply:
Yeah, got it. The name is quite confusing :-)